[FLINK-2432] Custom serializer support

zentol opened the pull request:
Users can now  use custom serializers in the Python API.
The registration is done using env.register_custom_type(<type class>,<serializer instance>, <deserializer instnace>)
Objects serialized this way are exist on a java side in serialized form, stored in a CustomTypeWrapper object.

rmetzger commented on flink-staging/flink-language-binding/flink-language-binding-generic/src/main/java/org/apache/flink/languagebinding/api/java/common/streaming/Sender.java:
```diff
@@ -253,6 +257,18 @@ public ByteBuffer serialize(T value) {
+		@Override
+		public void serializeInternal(CustomTypeWrapper value) {
+			byte[] bytes = value.getData();
+			buffer = ByteBuffer.allocate(bytes.length);
```
I guess you can use `ByteBuffer.wrap()` to avoid allocating an additional byte array for every custom type?

zentol replied:
Good point!

mxm commented on flink-staging/flink-language-binding/flink-python/src/test/python/org/apache/flink/languagebinding/api/python/flink/test/test_custom.py:
```diff
@@ -0,0 +1,76 @@
+# ###############################################################################
```
The name of this file could be a bit more explanatory like: "test_custom_type_serializer"

zentol replied:
Actually, I'm just gonna move this code into the test_main file, I was going to do that in my next test-centric PR anyway.

mxm replied:
Ok, it should reside there with a descriptive name as well.

zentol commented:
I've addressed the mentioned issues, and added a CUSTOM type constant, so that users don't have to constantly create new instances of their classes just to pass as a type parameter.

StephanEwen commented:
What is the status of this? Is this ready from your perspective, @zentol ?

zentol commented:
yep its good do go.

StephanEwen commented:
@zentol I am not sure anyone has the capacity to dig into this right now.

If you feel like this is a good change, I'd say go ahead and merge it... (you are the most knowledgeable person about the Python API anyways).

zentol commented:
WIll merge this once Travis finishes.
